src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs

```diff
@@ -296,38 +264,106 @@ public void UnregisterPrimarySolution(SolutionId solutionId, bool synchronousShu
+            }
+
+            ReleaseStorage(storage, synchronousShutdown);
+        }
+
+        private static void ReleaseStorage(AbstractPersistentStorage storage, bool synchronousShutdown)
+        {
+            if (storage == null)
+            {
+                return;
+            }
+
+            if (synchronousShutdown)
+            {
+                // dispose storage outside of the lock
+                storage.Dispose();
+            }
+            else
+            {
+                // make it to shutdown asynchronously
+                Task.Run(() => storage.Dispose());
+            }
+        }
+
+        private class PrimaryStorageInfo
+        {
+            public SolutionId SolutionId { get; private set; }
+
+            public AbstractPersistentStorage Storage { get; private set; }
+
+            public bool TryGetStorage(
+                string solutionFilePath,
+                string workingFolderPath,
+                out AbstractPersistentStorage storage)
+            {
+                storage = null;
+
+                if (Storage == null)
                 {
-                    // primary solution is never registered or already unregistered
-                    Contract.ThrowIfTrue(_primarySolutionStorage != null);
-                    return;
+                    return false;
+                }
+
+                if (Storage.SolutionFilePath != solutionFilePath ||
+                    Storage.WorkingFolderPath != workingFolderPath)
+                {
+                    return false;
                 }
 
-                Contract.ThrowIfFalse(_primarySolutionId == solutionId);
+                storage = Storage;
+                storage.AddRefUnsafe();
 
-                _primarySolutionId = null;
-                if (_primarySolutionStorage == null)
+                return true;
+            }
+
+            public void EnsureStorage(Solution solution, AbstractPersistentStorage storage)
+            {
+                if (Storage != null || solution.Id != SolutionId)
                 {
-                    // primary solution is registered but no C#/VB project was added
                     return;
                 }
 
-                storage = _primarySolutionStorage;
-                _primarySolutionStorage = null;
+                // hold onto the primary solution when it is used the first time.
+                Storage = storage;
+                storage.AddRefUnsafe();
             }
 
-            if (storage != null)
+            public void RegisterPrimarySolution(SolutionId solutionId)
             {
-                if (synchronousShutdown)
+                Contract.ThrowIfNull(solutionId);
+
+                if (SolutionId == solutionId)
                 {
-                    // dispose storage outside of the lock
-                    storage.Dispose();
+                    // this can happen if user opened solution without creating actual file
```

Comment:
@sharwell this (if solutionid == ) block is the actual fix. all other is just clean up.